src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
 func (a *metricsAppender) NextMetric() {
 	a.tags.names = a.tags.names[:0]
 	a.tags.values = a.tags.values[:0]
+
+	// Move the inuse encoders to cached as we should be done with using them.
+	a.moveInuseEncoders()
 }
 
 func (a *metricsAppender) Finalize() {
 	// Return to pool.
 	a.pool.Put(a)
 }
 
+func (a *metricsAppender) moveInuseEncoders() {
+	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
+	for i := range a.inuseEncoders {
+		a.inuseEncoders[i] = nil
+	}
+	a.inuseEncoders = a.inuseEncoders[:0]
+}
+
+func (a *metricsAppender) tagEncoder() serialize.TagEncoder {
+	// Take an encoder from the cached encoder list, if not present get one
+	// from the pool. Add the returned encoder to the used list.
+	var tagEncoder serialize.TagEncoder
+	if len(a.cachedEncoders) == 0 {
+		tagEncoder = a.tagEncoderPool.Get()
+	} else {
+		tagEncoder = a.cachedEncoders[0]
+		a.cachedEncoders = a.cachedEncoders[1:]
```

Comment:
Perhaps better to take from the tail, nil out the tail entry index, then shrink so the slice doesn't endlessly grow by move to the right always?

Reply:
true yeah, that's better.